### PR TITLE
TSServer Types2.0

### PR DESCRIPTION
**Pull request by jramsay**
TSServer Types2.0
- replacing TSD with Types 2,0
Note:  tests to follow shortly

### CODE REVIEW

File: src/server/typingsInstaller/nodeTypingsInstaller.ts [hunk 1]
```diff
@@ -106,40 +89,69 @@ namespace ts.server.typingsInstaller {
...
+                this.exec(command, { cwd: cachePath }, (err, stdout, stderr) => {
+                    execInstallCmdCount++;
+                    if (this.log.isEnabled()) {
+                        this.log.writeLine(`npm install @types ${id} stdout: ${stdout}`);
+                        this.log.writeLine(`npm install @types ${id} stderr: ${stderr}`);
+                    }
+                    if (stdout !== "") {
+                        filteredTypings.push(typing);
+                    }
+                    if (execInstallCmdCount >= typingsToInstall.length) {
```
Comment (vladima) [inline src/server/typingsInstaller/nodeTypingsInstaller.ts]: why `>=` and not just `===`?

File: src/server/typingsInstaller/nodeTypingsInstaller.ts [hunk 2]
```diff
@@ -106,40 +89,69 @@ namespace ts.server.typingsInstaller {
...
-        protected runTsd(cachePath: string, typingsToInstall: string[], postInstallAction: (installedTypings: string[]) => void): void {
-            const id = this.tsdRunCount;
-            this.tsdRunCount++;
-            const tsdPath = combinePaths(this.npmBinPath, "tsd");
-            const command = `${tsdPath} install ${typingsToInstall.join(" ")} -ros`;
-            if (this.log.isEnabled()) {
-                this.log.writeLine(`Running tsd ${id}, command '${command}'. cache path '${cachePath}'`);
-            }
-            this.exec(command, { cwd: cachePath }, (err, stdout, stderr) => {
+        protected runInstall(cachePath: string, typingsToInstall: string[], postInstallAction: (installedTypings: string[]) => void): void {
```
Comment (vladima) [inline src/server/typingsInstaller/nodeTypingsInstaller.ts]: I would probably write it a bit differently to reduce nesting and code duplication, smth like below (warning, typed directly in browser, may contain ~~peanuts~~ errors)

```ts
private execAsync(prefix: string, command: string, cwd: string, requestId: number, cb: (err: Error, stdout: string, stderr: string) => void) {
    if (this.log.isEnabled()) {
        this.log.writeLine(`#${requestId} running command '${command}'.`);
    }
    this.exec(command, { cwd }, (err, stdout, stderr) => {
        if (this.log.isEnabled()) {
            this.log.writeLine(`${prefix} #${requestId} stdout: ${stdout}`);
            this.log.writeLine(`${prefix} #${requestId} stderr: ${stderr}`);
        }
        cb(err, stdout, stderr);
    });
}

protected runInstall(cachePath: string, typingsToInstall: string[], postInstallAction: (installedTypings: string[]) => void): void {
    const id = this.installRunCount;
    this.installRunCount++;
    let execInstallCmdCount = 0;
    let filteredTypings: string[] = [];
    
    for (const typing of typingsToInstall) {
        const command = `npm view @types/${typing} --silent name`;
        this.execAsync("npm view @types", command, cachePath, id, (err, stdout, stderr) => {
            if (stdout) {
                filteredTypings.push(stdout);
            }
            execInstallCmdCount++;
            if (execInstallCmdCount === typingsToInstall.length) {
                installFilteredTypings(this, filteredTypings);
            }
        })
    }

    function installFilteredTypings(self: NodeTypingsInstaller, filteredTypings: string[]) {
        const command = `npm install ${filteredTypings.map(t => "@types/" + t).join(" ")} --save-dev`;
        self.execAsync("npm install", command, cachePath, id, (err, stdout, stderr) => {
            if (stdout) {
                reportInstalledTypings(self)
            }
        })
    }

    function reportInstalledTypings(self: NodeTypingsInstaller) {
        const command = "npm ls -json";
        self.execAsync("npm ls", command, cachePath, id, (err, stdout, stderr) => {
            let installedTypings: string[];
            try {
                const response = JSON.parse(stdout);
                if (response.dependencies) {
                    installedTypings = getOwnKeys(response.dependencies);
                }
            }
            catch (e) {
                self.log.writeLine(`Error parsing installed @types dependencies. Error details: ${e.message}`);
            }
            postInstallAction(installedTypings || []);
        });
    }
```

File: src/server/typingsInstaller/nodeTypingsInstaller.ts [hunk 3]
```diff
@@ -106,40 +89,69 @@ namespace ts.server.typingsInstaller {
...
+                                this.log.writeLine(`npm install @types ${id} stdout: ${stdout}`);
+                                this.log.writeLine(`npm install @types ${id} stderr: ${stderr}`);
+                            }
+                            if (stdout === "") {
+                                return;
+                            }
+                            const command = "npm ls -json";
+                            this.exec(command, { cwd: cachePath }, (err, stdout, stderr) => {
+                                if (this.log.isEnabled()) {
+                                    this.log.writeLine(`npm ls -json ${id} stdout: ${stdout}`);
```
Comment (vladima) [inline src/server/typingsInstaller/nodeTypingsInstaller.ts]: also add `--depth = 0`

File: src/server/typingsInstaller/nodeTypingsInstaller.ts (hunk 1 above)
Comment (jramsay) [inline src/server/typingsInstaller/nodeTypingsInstaller.ts]: === is more precise here as it should never be > - changed

File: src/server/typingsInstaller/nodeTypingsInstaller.ts (hunk 3 above)
Comment (jramsay) [inline src/server/typingsInstaller/nodeTypingsInstaller.ts]: --depth=0 doesn't have any impact on -json output

File: src/server/typingsInstaller/nodeTypingsInstaller.ts (hunk 2 above)
Comment (jramsay) [inline src/server/typingsInstaller/nodeTypingsInstaller.ts]: Thanks! much cleaner - will make this change - EpiPen is ready :)

**Comment by jramsay** [comment]
@vladima:  will add more tests - the current iteration fixes up the ones we already have

File: src/harness/unittests/typingsInstaller.ts [hunk 4]
```diff
@@ -4,7 +4,8 @@
 
 namespace ts.projectSystem {
     describe("typings installer", () => {
-        it("configured projects (tsd installed) 1", () => {
+        it("configured projects (typings installed) 1", () => {
+            debugger;
```
Comment (vladima) [inline src/harness/unittests/typingsInstaller.ts]: Drop `debugger`

Comment (jramsay) [inline src/harness/unittests/typingsInstaller.ts]: done :)

**Pull request merged**